File: sysplant/templates/iterators/syswhispers3.c
PVOID SC_Address(PVOID NtApiAddress)
{
    DWORD searchLimit = 512;
    PVOID SyscallAddress;

#ifdef _WIN64
    // If the process is 64-bit on a 64-bit OS, we need to search for syscall
    BYTE syscall_code[] = {0x0f, 0x05, 0xc3};
    ULONG distance_to_syscall = 0x12;
#else
    // If the process is 32-bit on a 32-bit OS, we need to search for sysenter
    BYTE syscall_code[] = {0x0f, 0x34, 0xc3};
    ULONG distance_to_syscall = 0x0f;
#endif

#ifdef _M_IX86
    // If the process is 32-bit on a 64-bit OS, we need to jump to WOW32Reserved
    if (local_is_wow64())
    {
        return NULL;
    }
#endif

    // we don't really care if there is a 'jmp' between
    // NtApiAddress and the 'syscall; ret' instructions
    SyscallAddress = SPT_RVA2VA(PVOID, NtApiAddress, distance_to_syscall);

    if (!memcmp((PVOID)syscall_code, SyscallAddress, sizeof(syscall_code)))
    {
        return SyscallAddress;
    }

    // the 'syscall; ret' intructions have not been found,
    // we will try to use one near it, similarly to HalosGate

    for (ULONG32 num_jumps = 1; num_jumps < searchLimit; num_jumps++)
    {
        // let's try with an Nt* API below our syscall
        SyscallAddress = SPT_RVA2VA(
            PVOID,
            NtApiAddress,
            distance_to_syscall + num_jumps * 0x20);
        if (!memcmp((PVOID)syscall_code, SyscallAddress, sizeof(syscall_code)))
        {
            return SyscallAddress;
        }

        // let's try with an Nt* API above our syscall
        SyscallAddress = SPT_RVA2VA(
            PVOID,
            NtApiAddress,
            distance_to_syscall - num_jumps * 0x20);
        if (!memcmp((PVOID)syscall_code, SyscallAddress, sizeof(syscall_code)))
        {
            return SyscallAddress;
        }
    }

    return NULL;
}

// Parse Export Directory and lookup syscall by name (start with Nt and not Ntdll), sort addresses to retrieve syscall number https://github.com/crummie5/FreshyCalls
BOOL SPT_PopulateSyscallList()
{
    // Return early if the list is already populated.
    if (SPT_SyscallList.Count) return TRUE;

    #ifdef _WIN64
        PSPT_PEB Peb = (PSPT_PEB)__readgsqword(0x60);
    #else
        PSPT_PEB Peb = (PSPT_PEB)__readfsdword(0x30);
    #endif
    PSPT_PEB_LDR_DATA Ldr = Peb->Ldr;
    PIMAGE_EXPORT_DIRECTORY ExportDirectory = NULL;
    PVOID DllBase = NULL;

    // Get the DllBase address of NTDLL.dll. NTDLL is not guaranteed to be the second
    // in the list, so it's safer to loop through the full list and find it.
    PSPT_LDR_DATA_TABLE_ENTRY LdrEntry;
    for (LdrEntry = (PSPT_LDR_DATA_TABLE_ENTRY)Ldr->Reserved2[1]; LdrEntry->DllBase != NULL; LdrEntry = (PSPT_LDR_DATA_TABLE_ENTRY)LdrEntry->Reserved1[0])
    {
        DllBase = LdrEntry->DllBase;
        PIMAGE_DOS_HEADER DosHeader = (PIMAGE_DOS_HEADER)DllBase;
        PIMAGE_NT_HEADERS NtHeaders = SPT_RVA2VA(PIMAGE_NT_HEADERS, DllBase, DosHeader->e_lfanew);
        PIMAGE_DATA_DIRECTORY DataDirectory = (PIMAGE_DATA_DIRECTORY)NtHeaders->OptionalHeader.DataDirectory;
        DWORD VirtualAddress = DataDirectory[IMAGE_DIRECTORY_ENTRY_EXPORT].VirtualAddress;
        if (VirtualAddress == 0) continue;

        ExportDirectory = (PIMAGE_EXPORT_DIRECTORY)SPT_RVA2VA(ULONG_PTR, DllBase, VirtualAddress);

        // If this is NTDLL.dll, exit loop.
        PCHAR DllName = SPT_RVA2VA(PCHAR, DllBase, ExportDirectory->Name);

        if ((*(ULONG*)DllName | 0x20202020) != 0x6c64746e) continue;
        if ((*(ULONG*)(DllName + 4) | 0x20202020) == 0x6c642e6c) break;
    }

    if (!ExportDirectory) return FALSE;

    DWORD NumberOfNames = ExportDirectory->NumberOfNames;
    PDWORD Functions = SPT_RVA2VA(PDWORD, DllBase, ExportDirectory->AddressOfFunctions);
    PDWORD Names = SPT_RVA2VA(PDWORD, DllBase, ExportDirectory->AddressOfNames);
    PWORD Ordinals = SPT_RVA2VA(PWORD, DllBase, ExportDirectory->AddressOfNameOrdinals);

    // Populate SPT_SyscallList with unsorted Zw* entries.
    DWORD i = 0;
    WORD padding = 0x0;
    PSPT_SYSCALL_ENTRY Entries = SPT_SyscallList.Entries;
    do
    {
        PCHAR FunctionName = SPT_RVA2VA(PCHAR, DllBase, Names[NumberOfNames - 1]);
        PVOID FunctionAddress = SPT_RVA2VA(PVOID, DllBase, Functions[Ordinals[NumberOfNames - 1]]);

        // Is this a system call?
        if (*(USHORT*)FunctionName == 0x775a)
        {
            Entries[i].Hash = SPT_HashSyscallName(FunctionName);
            Entries[i].Address = FunctionAddress;

            // All syscall stubs are identical for a Windows version
            if (padding == 0x0) {
                padding = SPT_DetectPadding(FunctionAddress);
            }

            Entries[i].SyscallAddress = SPT_RVA2VA(PVOID, Entries[i].Address, padding);

            i++;
            if (i == SPT_MAX_ENTRIES) break;
        }
    } while (--NumberOfNames);

    // Save total number of system calls found.
    SPT_SyscallList.Count = i;

    // Sort the list by address in ascending order.
    for (DWORD i = 0; i < SPT_SyscallList.Count - 1; i++)
    {
        for (DWORD j = 0; j < SPT_SyscallList.Count - i - 1; j++)
        {
            if (Entries[j].Address > Entries[j + 1].Address)
            {
                // Swap entries.
                SPT_SYSCALL_ENTRY TempEntry;

                TempEntry.Hash = Entries[j].Hash;
                TempEntry.Address = Entries[j].Address;
                TempEntry.SyscallAddress = Entries[j].SyscallAddress;

                Entries[j].Hash = Entries[j + 1].Hash;
                Entries[j].Address = Entries[j + 1].Address;
                Entries[j].SyscallAddress = Entries[j + 1].SyscallAddress;

                Entries[j + 1].Hash = TempEntry.Hash;
                Entries[j + 1].Address = TempEntry.Address;
                Entries[j + 1].SyscallAddress = TempEntry.SyscallAddress;
            }
        }
    }

    return TRUE;
}

File: sysplant/templates/iterators/syswhispers3.rs
unsafe fn sc_address(nt_api_address: PVOID) -> PVOID {
    let search_limit: u32 = 512;

    #[cfg(target_arch = "x86_64")]
    let syscall_code: [u8; 3] = [0x0f, 0x05, 0xc3];
    #[cfg(target_arch = "x86_64")]
    let distance_to_syscall: u32 = 0x12;

    #[cfg(target_arch = "x86")]
    let syscall_code: [u8; 3] = [0x0f, 0x34, 0xc3];
    #[cfg(target_arch = "x86")]
    let distance_to_syscall: u32 = 0x0f;

    // we don't really care if there is a 'jmp' between
    // NtApiAddress and the 'syscall; ret' instructions
    let syscall_address = SPT_RVA2VA!(PVOID, nt_api_address, distance_to_syscall);

    if core::slice::from_raw_parts(syscall_address as *const u8, 3) == syscall_code {
        return syscall_address;
    }

    // the 'syscall; ret' instructions have not been found,
    // we will try to use one near it, similarly to HalosGate
    for num_jumps in 1..search_limit {
        // let's try with an Nt* API below our syscall
        let addr_below = SPT_RVA2VA!(
            PVOID,
            nt_api_address,
            distance_to_syscall + num_jumps * 0x20
        );
        if core::slice::from_raw_parts(addr_below as *const u8, 3) == syscall_code {
            return addr_below;
        }

        // let's try with an Nt* API above our syscall
        let addr_above = SPT_RVA2VA!(
            PVOID,
            nt_api_address,
            distance_to_syscall.wrapping_sub(num_jumps * 0x20)
        );
        if core::slice::from_raw_parts(addr_above as *const u8, 3) == syscall_code {
            return addr_above;
        }
    }

    ptr::null_mut()
}

// Parse Export Directory and lookup syscall by name (start with Nt and not Ntdll),
// sort addresses to retrieve syscall number (SysWhispers3 variant)
// https://github.com/crummie5/FreshyCalls
#[unsafe(no_mangle)]
unsafe fn SPT_PopulateSyscallList() -> BOOL {
    if SPT_SyscallList.Count != 0 {
        return TRUE;
    }

    let peb = get_peb();
    let ldr = (*peb).Ldr;
    let mut export_directory: *mut IMAGE_EXPORT_DIRECTORY = ptr::null_mut();
    let mut dll_base: PVOID = ptr::null_mut();

    let mut ldr_entry = (*ldr).InMemoryOrderModuleList.Flink as *mut SPT_LDR_DATA_TABLE_ENTRY;
    while !(*ldr_entry).DllBase.is_null() {
        dll_base = (*ldr_entry).DllBase;
        let dos_header = dll_base as *const IMAGE_DOS_HEADER;
        let nt_headers = SPT_RVA2VA!(*const IMAGE_NT_HEADERS64, dll_base, (*dos_header).e_lfanew);
        let data_directory = (*nt_headers).OptionalHeader.DataDirectory.as_ptr();
        let va = (*data_directory.add(IMAGE_DIRECTORY_ENTRY_EXPORT)).VirtualAddress;
        if va == 0 {
            ldr_entry = (*ldr_entry).InMemoryOrderLinks.Flink as *mut SPT_LDR_DATA_TABLE_ENTRY;
            continue;
        }

        export_directory = SPT_RVA2VA!(*mut IMAGE_EXPORT_DIRECTORY, dll_base, va);

        let dll_name = SPT_RVA2VA!(*const u8, dll_base, (*export_directory).Name);
        if (*(dll_name as *const u32) | 0x20202020) != 0x6c64746e {
            ldr_entry = (*ldr_entry).InMemoryOrderLinks.Flink as *mut SPT_LDR_DATA_TABLE_ENTRY;
            continue;
        }
        if (*(dll_name.add(4) as *const u32) | 0x20202020) == 0x6c642e6c {
            break;
        }
        ldr_entry = (*ldr_entry).InMemoryOrderLinks.Flink as *mut SPT_LDR_DATA_TABLE_ENTRY;
    }

    if export_directory.is_null() {
        return FALSE;
    }

    let mut number_of_names = (*export_directory).NumberOfNames;
    let functions = SPT_RVA2VA!(*const DWORD, dll_base, (*export_directory).AddressOfFunctions);
    let names = SPT_RVA2VA!(*const DWORD, dll_base, (*export_directory).AddressOfNames);
    let ordinals = SPT_RVA2VA!(*const WORD, dll_base, (*export_directory).AddressOfNameOrdinals);

    let mut i: u32 = 0;
    let mut padding: u16 = 0;
    let entries = SPT_SyscallList.Entries.as_mut_ptr();
    while number_of_names > 0 {
        number_of_names -= 1;
        let function_name = SPT_RVA2VA!(*const i8, dll_base, *names.add(number_of_names as usize));
        let function_address = SPT_RVA2VA!(PVOID, dll_base, *functions.add(*ordinals.add(number_of_names as usize) as usize));

        // Is this a system call? "Zw" == 0x775a
        if *(function_name as *const u16) == 0x775a {
            (*entries.add(i as usize)).Hash = SPT_HashSyscallName(function_name);
            (*entries.add(i as usize)).Address = function_address;

            if padding == 0 {
                padding = SPT_DetectPadding(function_address);
            }

            (*entries.add(i as usize)).SyscallAddress =
                SPT_RVA2VA!(PVOID, (*entries.add(i as usize)).Address, padding);

            i += 1;
            if i == SPT_MAX_ENTRIES as u32 {
                break;
            }
        }
    }

    SPT_SyscallList.Count = i;

    // Sort the list by address in ascending order (bubble sort).
    for i in 0..(SPT_SyscallList.Count as usize).saturating_sub(1) {
        for j in 0..(SPT_SyscallList.Count as usize - i - 1) {
            if (*entries.add(j)).Address as usize > (*entries.add(j + 1)).Address as usize {
                let temp_hash = (*entries.add(j)).Hash;
                let temp_addr = (*entries.add(j)).Address;
                let temp_sys = (*entries.add(j)).SyscallAddress;

                (*entries.add(j)).Hash = (*entries.add(j + 1)).Hash;
                (*entries.add(j)).Address = (*entries.add(j + 1)).Address;
                (*entries.add(j)).SyscallAddress = (*entries.add(j + 1)).SyscallAddress;

                (*entries.add(j + 1)).Hash = temp_hash;
                (*entries.add(j + 1)).Address = temp_addr;
                (*entries.add(j + 1)).SyscallAddress = temp_sys;
            }
        }
    }

    TRUE
}
